app/routines/confirm_by_pin.rb

```diff
@@ -0,0 +1,41 @@
+class ConfirmByPin
+  lev_routine
+
+  uses_routine ConfirmContactInfo
+
+  MAX_PIN_FAILURES = 5
+
+  def self.sequential_failure_for(contact_info)
+    SequentialFailure.confirm_by_pin
+                     .find_or_initialize_by_reference(contact_info.value).tap do |sf|
+      sf.num_failures_allowed = MAX_PIN_FAILURES
+    end
+  end
+
+  protected
+
+  def exec(contact_info:, pin:)
+    return if contact_info.confirmed?
+
+    sequential_failure = self.class.sequential_failure_for(contact_info)
+
+    if !sequential_failure.attempts_remaining?
+      fatal_error(code: :no_pin_confirmation_attempts_remaining)
+    else
+      if contact_info.confirmation_pin == pin
+        after_transaction do
+          sequential_failure.reset!
+        end
+
+        run(ConfirmContactInfo, contact_info)
+      else
+        after_transaction do
+          sequential_failure.increment!
+        end
+
```

Comment:
I'm probably missing something, but why are call `.reset!` and `.increment!` in an `after_transaction`   Doesn't that mean that those methods may fail silently without causing the lev routine to fail?  Maybe we want that behavior for some reason though.